src/nsolid/nsolid_heap_snapshot.cc

```diff
@@ -83,6 +83,71 @@ int NSolidHeapSnapshot::StopTrackingHeapObjects(SharedEnvInst envinst) {
   return er;
 }
 
+int NSolidHeapSnapshot::StopTrackingHeapObjectsSync(SharedEnvInst envinst) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&in_progress_heap_snapshots_);
+  auto it = threads_running_snapshots_.find(thread_id);
+  // Make sure there is a snapshot running
+  if (it == threads_running_snapshots_.end())
+    return UV_ENOENT;
+
+  HeapSnapshotStor& stor = it->second;
+  // If this condition is reached. This was called by EnvList::RemoveEnv.
+  // It wants to stop any pending snapshot w/ tracking heap object.
+  if (!stor.is_tracking_heapobjects_) {
+    // If no pending trackers, just do nothing
+    // There are not peding snapshots with trackers
+    return UV_ENOENT;
+  }
+
+  v8::Isolate* isolate = envinst->isolate();
+
+  v8::HeapProfiler* profiler = isolate->GetHeapProfiler();
+  ASSERT_NOT_NULL(profiler);
+
+  v8::HandleScope scope(isolate);
+
+  const v8::HeapSnapshot* snapshot = profiler->TakeHeapSnapshot();
+  if (snapshot == nullptr) {
+    stor.cb(heap_profiler::HEAP_SNAPSHOT_FAILURE,
+            std::string(),
+            stor.data.get());
+  } else {
+    std::string snapshot_str;
+    DataOutputStream<uint64_t, v8::HeapSnapshot> stream(&snapshot_str,
+                                                        snapshot,
+                                                        &thread_id);
+
```

Comment:
Why can't we use the "streaming" version of the DataOutputStream here. Both are sync so we should be fine

Reply:
This comment as discussed is for the async stoped. Fixed in https://github.com/nodesource/nsolid/commit/37d63c035e309c433c4af1c02774d4aec1f79ee5